File: examples/tts_example.py
"""
Примеры использования TTS API

Этот файл содержит практические примеры работы с Voice API EDIS.
"""
import asyncio
import httpx
from pathlib import Path


# Конфигурация
API_URL = "http://localhost:8000"
API_KEY = "your-secret-key-here"


async def example_1_simple_tts():
    """
    Пример 1: Простой синтез речи
    
    Самый базовый пример - преобразование текста в речь.
    """
    print("=== Пример 1: Простой TTS ===\n")
    
    async with httpx.AsyncClient() as client:
        response = await client.post(
            f"{API_URL}/api/v1/voice/tts",
            json={
                "text": "Привет! Это пример синтеза речи.",
                "language": "ru",
                "speed": 1.0
            },
            headers={"X-API-Key": API_KEY}
        )
        
        if response.status_code == 200:
            data = response.json()
            print(f"✅ Аудио создано: {data['audio_url']}")
            print(f"   Длительность: {data['duration']:.2f} сек")
        else:
            print(f"❌ Ошибка: {response.status_code}")
            print(f"   {response.text}")


async def example_2_create_voice():
    """
    Пример 2: Создание нового голоса
    
    Создание голоса из референсного аудио файла.
    """
    print("\n=== Пример 2: Создание голоса ===\n")
    
    # Путь к вашему аудио файлу
    audio_path = Path("path/to/your/voice_sample.wav")
    
    if not audio_path.exists():
        print("⚠️  Создайте аудио файл для примера")
        return
    
    async with httpx.AsyncClient() as client:
        with open(audio_path, "rb") as audio_file:
            response = await client.post(
                f"{API_URL}/api/v1/voice/voices",
                files={"audio": audio_file},
                data={
                    "name": "Мой голос",
                    "reference_text": "Это референсный текст для моего голоса",
                    "description": "Голос созданный из примера",
                    "language": "ru"
                },
                headers={"X-API-Key": API_KEY}
            )
        
        if response.status_code == 200:
            voice = response.json()
            print(f"✅ Голос создан!")
            print(f"   ID: {voice['id']}")
            print(f"   Название: {voice['name']}")
            print(f"   Язык: {voice['language']}")
            return voice['id']
        else:
            print(f"❌ Ошибка: {response.status_code}")


async def example_3_list_voices():
    """
    Пример 3: Получение списка голосов
    
    Просмотр всех доступных голосов в библиотеке.
    """
    print("\n=== Пример 3: Список голосов ===\n")
    
    async with httpx.AsyncClient() as client:
        response = await client.get(
            f"{API_URL}/api/v1/voice/voices",
            headers={"X-API-Key": API_KEY}
        )
        
        if response.status_code == 200:
            data = response.json()
            print(f"📚 Всего голосов: {data['total']}\n")
            
            for voice in data['voices']:
                status = "✅ Обучен" if voice['is_trained'] else "⏳ Не обучен"
                print(f"🎤 {voice['name']}")
                print(f"   ID: {voice['id']}")
                print(f"   Язык: {voice['language']}")
                print(f"   Статус: {status}")
                print(f"   Сэмплов: {voice['sample_count']}")
                print()
        else:
            print(f"❌ Ошибка: {response.status_code}")


async def example_4_add_training_samples():
    """
    Пример 4: Добавление обучающих сэмплов
    
    Добавление нескольких аудио сэмплов для обучения голоса.
    """
    print("\n=== Пример 4: Добавление сэмплов ===\n")
    
    voice_id = "your-voice-id-here"  # Замените на реальный ID
    
    # Список сэмплов для добавления
    samples = [
        ("sample1.wav", "Первый обучающий сэмпл"),
        ("sample2.wav", "Второй обучающий сэмпл"),
        ("sample3.wav", "Третий обучающий сэмпл"),
    ]
    
    async with httpx.AsyncClient() as client:
        for audio_file, text in samples:
            audio_path = Path(f"samples/{audio_file}")
            
            if not audio_path.exists():
                print(f"⚠️  Файл {audio_file} не найден, пропускаем")
                continue
            
            with open(audio_path, "rb") as f:
                response = await client.post(
                    f"{API_URL}/api/v1/voice/voices/{voice_id}/samples",
                    files={"audio": f},
                    data={"text": text},
                    headers={"X-API-Key": API_KEY}
                )
            
            if response.status_code == 200:
                sample = response.json()
                print(f"✅ Сэмпл добавлен: {sample['id']}")
                print(f"   Текст: {text}")
                print(f"   Длительность: {sample['duration']:.2f} сек\n")
            else:
                print(f"❌ Ошибка добавления {audio_file}: {response.status_code}\n")


async def example_5_train_voice():
    """
    Пример 5: Обучение голоса
    
    Запуск обучения голоса с настройкой параметров.
    """
    print("\n=== Пример 5: Обучение голоса ===\n")
    
    voice_id = "your-voice-id-here"  # Замените на реальный ID
    
    async with httpx.AsyncClient(timeout=60.0) as client:
        # Запуск обучения
        response = await client.post(
            f"{API_URL}/api/v1/voice/voices/{voice_id}/train",
            json={
                "voice_id": voice_id,
                "epochs": 100,
                "batch_size": 4,
                "learning_rate": 0.0001,
                "save_interval": 10
            },
            headers={"X-API-Key": API_KEY}
        )
        
        if response.status_code == 200:
            status = response.json()
            print(f"✅ Обучение запущено!")
            print(f"   Статус: {status['status']}")
            print(f"   Эпох: {status['total_epochs']}")
            print()
            
            # Мониторинг прогресса
            print("📊 Мониторинг прогресса...\n")
            while True:
                await asyncio.sleep(5)  # Проверяем каждые 5 секунд
                
                status_response = await client.get(
                    f"{API_URL}/api/v1/voice/voices/{voice_id}/training-status",
                    headers={"X-API-Key": API_KEY}
                )
                
                if status_response.status_code == 200:
                    status = status_response.json()
                    
                    if status['status'] == 'completed':
                        print(f"✅ Обучение завершено!")
                        break
                    elif status['status'] == 'failed':
                        print(f"❌ Обучение провалилось: {status.get('error')}")
                        break
                    else:
                        progress = status['progress']
                        epoch = status['current_epoch']
                        total = status['total_epochs']
                        eta = status.get('eta_seconds', 0)
                        
                        print(f"⏳ Прогресс: {progress:.1f}% | Эпоха: {epoch}/{total} | ETA: {eta//60} мин")
        else:
            print(f"❌ Ошибка: {response.status_code}")
            print(f"   {response.text}")


async def example_6_test_voice():
    """
    Пример 6: Тестирование голоса
    
    Генерация тестового аудио с обученным голосом.
    """
    print("\n=== Пример 6: Тестирование голоса ===\n")
    
    voice_id = "your-voice-id-here"  # Замените на реальный ID
    
    test_texts = [
        "Привет! Это тестовое сообщение.",
        "Как дела? Надеюсь, всё хорошо!",
        "Это пример синтеза речи с обученным голосом."
    ]
    
    async with httpx.AsyncClient() as client:
        for i, text in enumerate(test_texts, 1):
            response = await client.post(
                f"{API_URL}/api/v1/voice/voices/{voice_id}/test",
                json={
                    "voice_id": voice_id,
                    "test_text": text,
                    "language": "ru"
                },
                headers={"X-API-Key": API_KEY}
            )
            
            if response.status_code == 200:
                data = response.json()
                print(f"✅ Тест {i}: {data['audio_url']}")
                print(f"   Текст: {text}")
                print(f"   Длительность: {data['duration']:.2f} сек\n")
            else:
                print(f"❌ Ошибка теста {i}: {response.status_code}\n")


async def example_7_speech_to_text():
    """
    Пример 7: Распознавание речи (STT)
    
    Преобразование аудио в текст.
    """
    print("\n=== Пример 7: Speech-to-Text ===\n")
    
    audio_path = Path("path/to/audio_to_recognize.wav")
    
    if not audio_path.exists():
        print("⚠️  Создайте аудио файл для распознавания")
        return
    
    async with httpx.AsyncClient() as client:
        with open(audio_path, "rb") as audio_file:
            response = await client.post(
                f"{API_URL}/api/v1/voice/stt",
                files={"audio": audio_file},
                data={"language": "ru"},
                headers={"X-API-Key": API_KEY}
            )
        
        if response.status_code == 200:
            data = response.json()
            print(f"✅ Распознано:")
            print(f"   Текст: {data['text']}")
            print(f"   Уверенность: {data['confidence']:.2%}")
            print(f"   Язык: {data['language']}")
        else:
            print(f"❌ Ошибка: {response.status_code}")


async def example_8_batch_synthesis():
    """
    Пример 8: Пакетный синтез речи
    
    Генерация нескольких аудио файлов одновременно.
    """
    print("\n=== Пример 8: Пакетный синтез ===\n")
    
    texts = [
        "Первое сообщение для синтеза.",
        "Второе сообщение для синтеза.",
        "Третье сообщение для синтеза.",
        "Четвертое сообщение для синтеза.",
        "Пятое сообщение для синтеза."
    ]
    
    async with httpx.AsyncClient() as client:
        tasks = []
        for text in texts:
            task = client.post(
                f"{API_URL}/api/v1/voice/tts",
                json={"text": text, "language": "ru", "speed": 1.0},
                headers={"X-API-Key": API_KEY}
            )
            tasks.append(task)
        
        responses = await asyncio.gather(*tasks, return_exceptions=True)
        
        success_count = 0
        for i, response in enumerate(responses, 1):
            if isinstance(response, Exception):
                print(f"❌ Ошибка {i}: {response}")
            elif response.status_code == 200:
                data = response.json()
                print(f"✅ Аудио {i}: {data['audio_url']}")
                success_count += 1
            else:
                print(f"❌ Ошибка {i}: {response.status_code}")
        
        print(f"\n📊 Успешно: {success_count}/{len(texts)}")


async def main():
    """Запуск всех примеров"""
    print("🎤 Примеры использования Voice API EDIS\n")
    print("=" * 50)
    
    try:
        # Запускаем примеры по очереди
        await example_1_simple_tts()
        await example_3_list_voices()
        
        # Раскомментируйте нужные примеры:
        # await example_2_create_voice()
        # await example_4_add_training_samples()
        # await example_5_train_voice()
        # await example_6_test_voice()
        # await example_7_speech_to_text()
        # await example_8_batch_synthesis()
        
        print("\n" + "=" * 50)
        print("✅ Примеры завершены!")
        
    except Exception as e:
        print(f"\n❌ Ошибка: {e}")


if __name__ == "__main__":
    asyncio.run(main())

File: examples/tts_example.js
/**
 * Примеры использования Voice API на JavaScript/TypeScript
 * 
 * Этот файл содержит практические примеры работы с Voice API
 * из браузера или Node.js приложения.
 */

// Конфигурация
const API_URL = 'http://localhost:8000';
const API_KEY = 'your-secret-key-here';

// Базовый класс для работы с Voice API
class VoiceAPIClient {
  constructor(apiUrl = API_URL, apiKey = API_KEY) {
    this.apiUrl = apiUrl;
    this.apiKey = apiKey;
  }

  async request(endpoint, options = {}) {
    const url = `${this.apiUrl}${endpoint}`;
    const headers = {
      'X-API-Key': this.apiKey,
      ...options.headers
    };

    const response = await fetch(url, {
      ...options,
      headers
    });

    if (!response.ok) {
      throw new Error(`API Error: ${response.status} ${response.statusText}`);
    }

    return response.json();
  }
}

// ============================================
// Пример 1: Простой синтез речи
// ============================================
async function example1_simpleTTS() {
  console.log('=== Пример 1: Простой TTS ===\n');

  const client = new VoiceAPIClient();

  try {
    const result = await client.request('/api/v1/voice/tts', {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        text: 'Привет! Это пример синтеза речи.',
        language: 'ru',
        speed: 1.0
      })
    });

    console.log('✅ Аудио создано:', result.audio_url);
    console.log('   Длительность:', result.duration.toFixed(2), 'сек');

    // Воспроизведение в браузере
    if (typeof Audio !== 'undefined') {
      const audio = new Audio(`${API_URL}${result.audio_url}`);
      await audio.play();
    }

    return result;
  } catch (error) {
    console.error('❌ Ошибка:', error.message);
  }
}

// ============================================
// Пример 2: Создание голоса
// ============================================
async function example2_createVoice(audioFile, name, referenceText) {
  console.log('\n=== Пример 2: Создание голоса ===\n');

  const formData = new FormData();
  formData.append('audio', audioFile);
  formData.append('name', name);
  formData.append('reference_text', referenceText);
  formData.append('description', 'Голос созданный из примера');
  formData.append('language', 'ru');

  try {
    const response = await fetch(`${API_URL}/api/v1/voice/voices`, {
      method: 'POST',
      headers: { 'X-API-Key': API_KEY },
      body: formData
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }

    const voice = await response.json();
    console.log('✅ Голос создан!');
    console.log('   ID:', voice.id);
    console.log('   Название:', voice.name);
    console.log('   Язык:', voice.language);

    return voice;
  } catch (error) {
    console.error('❌ Ошибка:', error.message);
  }
}

// ============================================
// Пример 3: Получение списка голосов
// ============================================
async function example3_listVoices() {
  console.log('\n=== Пример 3: Список голосов ===\n');

  const client = new VoiceAPIClient();

  try {
    const data = await client.request('/api/v1/voice/voices');

    console.log(`📚 Всего голосов: ${data.total}\n`);

    data.voices.forEach(voice => {
      const status = voice.is_trained ? '✅ Обучен' : '⏳ Не обучен';
      console.log(`🎤 ${voice.name}`);
      console.log(`   ID: ${voice.id}`);
      console.log(`   Язык: ${voice.language}`);
      console.log(`   Статус: ${status}`);
      console.log(`   Сэмплов: ${voice.sample_count}\n`);
    });

    return data.voices;
  } catch (error) {
    console.error('❌ Ошибка:', error.message);
  }
}

// ============================================
// Пример 4: Добавление обучающего сэмпла
// ============================================
async function example4_addTrainingSample(voiceId, audioFile, text) {
  console.log('\n=== Пример 4: Добавление сэмпла ===\n');

  const formData = new FormData();
  formData.append('audio', audioFile);
  formData.append('text', text);

  try {
    const response = await fetch(
      `${API_URL}/api/v1/voice/voices/${voiceId}/samples`,
      {
        method: 'POST',
        headers: { 'X-API-Key': API_KEY },
        body: formData
      }
    );

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }

    const sample = await response.json();
    console.log('✅ Сэмпл добавлен:', sample.id);
    console.log('   Текст:', text);
    console.log('   Длительность:', sample.duration.toFixed(2), 'сек');

    return sample;
  } catch (error) {
    console.error('❌ Ошибка:', error.message);
  }
}

// ============================================
// Пример 5: Обучение голоса с мониторингом
// ============================================
async function example5_trainVoice(voiceId, options = {}) {
  console.log('\n=== Пример 5: Обучение голоса ===\n');

  const client = new VoiceAPIClient();

  const trainingParams = {
    voice_id: voiceId,
    epochs: options.epochs || 100,
    batch_size: options.batchSize || 4,
    learning_rate: options.learningRate || 0.0001,
    save_interval: options.saveInterval || 10
  };

  try {
    // Запуск обучения
    const status = await client.request(`/api/v1/voice/voices/${voiceId}/train`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(trainingParams)
    });

    console.log('✅ Обучение запущено!');
    console.log('   Статус:', status.status);
    console.log('   Эпох:', status.total_epochs);
    console.log();

    // Мониторинг прогресса
    console.log('📊 Мониторинг прогресса...\n');

    return new Promise((resolve, reject) => {
      const checkProgress = async () => {
        try {
          const currentStatus = await client.request(
            `/api/v1/voice/voices/${voiceId}/training-status`
          );

          if (currentStatus.status === 'completed') {
            console.log('✅ Обучение завершено!');
            resolve(currentStatus);
            return;
          }

          if (currentStatus.status === 'failed') {
            console.error('❌ Обучение провалилось:', currentStatus.error);
            reject(new Error(currentStatus.error));
            return;
          }

          // Вывод прогресса
          const progress = currentStatus.progress.toFixed(1);
          const epoch = currentStatus.current_epoch;
          const total = currentStatus.total_epochs;
          const eta = Math.floor((currentStatus.eta_seconds || 0) / 60);

          console.log(`⏳ Прогресс: ${progress}% | Эпоха: ${epoch}/${total} | ETA: ${eta} мин`);

          // Проверяем снова через 5 секунд
          setTimeout(checkProgress, 5000);
        } catch (error) {
          reject(error);
        }
      };

      checkProgress();
    });
  } catch (error) {
    console.error('❌ Ошибка:', error.message);
    throw error;
  }
}

// ============================================
// Пример 6: Тестирование голоса
// ============================================
async function example6_testVoice(voiceId, testText) {
  console.log('\n=== Пример 6: Тестирование голоса ===\n');

  const client = new VoiceAPIClient();

  try {
    const result = await client.request(`/api/v1/voice/voices/${voiceId}/test`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        voice_id: voiceId,
        test_text: testText,
        language: 'ru'
      })
    });

    console.log('✅ Тестовое аудио создано:', result.audio_url);
    console.log('   Текст:', testText);
    console.log('   Длительность:', result.duration.toFixed(2), 'сек');

    // Воспроизведение в браузере
    if (typeof Audio !== 'undefined') {
      const audio = new Audio(`${API_URL}${result.audio_url}`);
      await audio.play();
    }

    return result;
  } catch (error) {
    console.error('❌ Ошибка:', error.message);
  }
}

// ============================================
// Пример 7: Speech-to-Text
// ============================================
async function example7_speechToText(audioFile, language = 'ru') {
  console.log('\n=== Пример 7: Speech-to-Text ===\n');

  const formData = new FormData();
  formData.append('audio', audioFile);
  formData.append('language', language);

  try {
    const response = await fetch(`${API_URL}/api/v1/voice/stt`, {
      method: 'POST',
      headers: { 'X-API-Key': API_KEY },
      body: formData
    });

    if (!response.ok) {
      throw new Error(`HTTP ${response.status}`);
    }

    const result = await response.json();
    console.log('✅ Распознано:');
    console.log('   Текст:', result.text);
    console.log('   Уверенность:', (result.confidence * 100).toFixed(1) + '%');
    console.log('   Язык:', result.language);

    return result;
  } catch (error) {
    console.error('❌ Ошибка:', error.message);
  }
}

// ============================================
// Пример 8: React Hook для TTS
// ============================================
/**
 * Пример React Hook для использования в компонентах
 * 
 * Usage:
 * ```tsx
 * const { speak, isPlaying, error } = useTTS();
 * 
 * <button onClick={() => speak('Привет!')}>
 *   {isPlaying ? 'Играет...' : 'Озвучить'}
 * </button>
 * ```
 */
function useTTS(apiUrl = API_URL, apiKey = API_KEY) {
  const [isPlaying, setIsPlaying] = React.useState(false);
  const [error, setError] = React.useState(null);
  const audioRef = React.useRef(null);

  const speak = React.useCallback(async (text, options = {}) => {
    setError(null);
    setIsPlaying(true);

    try {
      const response = await fetch(`${apiUrl}/api/v1/voice/tts`, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'X-API-Key': apiKey
        },
        body: JSON.stringify({
          text,
          language: options.language || 'ru',
          speed: options.speed || 1.0,
          voice_id: options.voiceId
        })
      });

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }

      const result = await response.json();

      // Создаем и воспроизводим аудио
      const audio = new Audio(`${apiUrl}${result.audio_url}`);
      audioRef.current = audio;

      audio.onended = () => setIsPlaying(false);
      audio.onerror = () => {
        setError('Ошибка воспроизведения');
        setIsPlaying(false);
      };

      await audio.play();
    } catch (err) {
      setError(err.message);
      setIsPlaying(false);
    }
  }, [apiUrl, apiKey]);

  const stop = React.useCallback(() => {
    if (audioRef.current) {
      audioRef.current.pause();
      audioRef.current = null;
      setIsPlaying(false);
    }
  }, []);

  return { speak, stop, isPlaying, error };
}

// ============================================
// Пример 9: Пакетная обработка
// ============================================
async function example9_batchProcessing(texts) {
  console.log('\n=== Пример 9: Пакетная обработка ===\n');

  const client = new VoiceAPIClient();

  try {
    const promises = texts.map(text =>
      client.request('/api/v1/voice/tts', {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify({ text, language: 'ru', speed: 1.0 })
      })
    );

    const results = await Promise.allSettled(promises);

    let successCount = 0;
    results.forEach((result, index) => {
      if (result.status === 'fulfilled') {
        console.log(`✅ Аудио ${index + 1}:`, result.value.audio_url);
        successCount++;
      } else {
        console.error(`❌ Ошибка ${index + 1}:`, result.reason.message);
      }
    });

    console.log(`\n📊 Успешно: ${successCount}/${texts.length}`);

    return results;
  } catch (error) {
    console.error('❌ Ошибка:', error.message);
  }
}

// ============================================
// Экспорт для использования в модулях
// ============================================
if (typeof module !== 'undefined' && module.exports) {
  module.exports = {
    VoiceAPIClient,
    example1_simpleTTS,
    example2_createVoice,
    example3_listVoices,
    example4_addTrainingSample,
    example5_trainVoice,
    example6_testVoice,
    example7_speechToText,
    useTTS,
    example9_batchProcessing
  };
}

// ============================================
// Запуск примеров (если файл запущен напрямую)
// ============================================
if (typeof window !== 'undefined') {
  console.log('🎤 Примеры Voice API доступны в консоли');
  console.log('Используйте: example1_simpleTTS(), example3_listVoices() и т.д.');
}
